src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -102,22 +104,49 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
   }
 
   private String getLink(String id, String query) {
-    String cse_request = CSE_URL + "?key=" + API_KEY + "&cx=" + id + "&q=" + query;
-    HttpRequest linkRequest =
-        HttpRequest.newBuilder()
-            .GET()
-            .uri(URI.create(cse_request))
-            .setHeader("User-Agent", "Java 11 HttpClient Bot")
-            .build();
-    HttpResponse<String> linkResponse;
+    String url = CSE_URL + "?key=" + API_KEY + "&cx=" + id + "&q=" + query;
+    CloseableHttpClient httpClient = HttpClients.createDefault();
+    CloseableHttpResponse response;
+    try {
+      response = httpClient.execute(new HttpGet(url));
+    } catch (IOException e) {
+      return null;
+    }
+    if (response.getStatusLine().getStatusCode() != 200) {
+      return null;
+    }
+    HttpEntity entity = response.getEntity();
+    if (entity == null) {
+      return null;
+    }
+    BufferedReader reader;
     try {
-      linkResponse = httpClient.send(linkRequest, HttpResponse.BodyHandlers.ofString());
-    } catch (IOException | InterruptedException ex) {
+      reader = new BufferedReader(new InputStreamReader(entity.getContent()));
+    } catch (IOException e) {
       return null;
     }
-    /* Parse JSON of CSE SRP to retrieve link from only the first search result */
-    JSONObject obj = new JSONObject(linkResponse.body());
-    String link = obj.getJSONArray("items").getJSONObject(0).getString("link");
-    return link;
+    StringBuilder responseBody = new StringBuilder();
+    String line;
+    try {
+      while ((line = reader.readLine()) != null) {
+        responseBody.append(line);
+      }
+      reader.close();
+    } catch (IOException e) {
+      return null;
+    }
+    JSONObject json = new JSONObject(responseBody.toString());
```

Comment:
So I think you can shorten this method by about a third by directly calling `JSONObject.parse()` on your InputStreamReader. No need to create a BufferedReader, or to read into a StringBuilder, etc. You can use a Java 7 try-with-resources idiom to simultaneously catch exceptions and close resources.